mono/metadata/android.c

```diff
@@ -0,0 +1,537 @@
+	if (!NetworkInterface_isUp)
+		g_error ("Fatal error: Could not find java.net.NetworkInterface.isUp() method!");
+
+	NetworkInterface_supportsMulticast = (*env)->GetMethodID (env, NetworkInterface_class, "supportsMulticast", "()Z");
+	if (!NetworkInterface_supportsMulticast)
+		g_error ("Fatal error: Could not find java.net.NetworkInterface.supportsMulticast() method!");
+
+	initialized = TRUE;
+}
+
+JNIEXPORT jint JNICALL
+JNI_OnLoad (JavaVM *vm, gpointer reserved)
+{
+	mono_jvm_initialize (vm);
+	return JNI_VERSION_1_6;
+}
+
+JNIEnv*
+mono_jvm_get_jnienv (void)
+{
+	JNIEnv *env;
+
+	g_assert (initialized);
+
+	(*jvm)->GetEnv (jvm, (void**)&env, JNI_VERSION_1_6);
+	if (env)
+		return env;
+
+	mono_thread_attach (mono_domain_get ());
+
+	(*jvm)->GetEnv (jvm, (void**)&env, JNI_VERSION_1_6);
+	if (env)
+		return env;
+
+	g_error ("%s: Fatal error: Could not create env", __func__);
+}
+
+struct BundledProperty {
+	gchar *name;
+	gchar *value;
+	gint   value_len;
+	struct BundledProperty *next;
+};
+
+static struct BundledProperty* bundled_properties;
+
+static struct BundledProperty*
+lookup_system_property (const gchar *name)
+{
+	struct BundledProperty *p = bundled_properties;
+	for ( ; p ; p = p->next)
+		if (strcmp (p->name, name) == 0)
+			return p;
+	return NULL;
+}
+
+void
+monodroid_add_system_property (const gchar *name, const gchar *value)
+{
+	gint name_len, value_len;
+
+	struct BundledProperty* p = lookup_system_property (name);
+	if (p) {
+		gchar *n = g_strdup (value);
+		g_free (p->value);
+		p->value      = n;
+		p->value_len  = strlen (p->value);
+		return;
+	}
+
+	name_len  = strlen (name);
+	value_len = strlen (value);
+
+	p = g_malloc0 (sizeof (struct BundledProperty) + name_len + 1);
+
+	p->name = ((char*) p) + sizeof (struct BundledProperty);
+	strncpy (p->name, name, name_len);
+	p->name [name_len] = '\0';
+
+	p->value      = g_strdup (value);
+	p->value_len  = value_len;
+
+	p->next             = bundled_properties;
+	bundled_properties  = p;
+}
+
+#ifndef ANDROID
+static void
+monodroid_strreplace (gchar *buffer, gchar old_char, gchar new_char)
+{
+	if (buffer == NULL)
+		return;
+	while (*buffer != '\0') {
+		if (*buffer == old_char)
+			*buffer = new_char;
+		buffer++;
+	}
+}
+
+static gint
+_monodroid__system_property_get (const gchar *name, gchar *sp_value, gsize sp_value_len)
+{
+	if (!name)
+		return -1;
+
+	g_assert (sp_value);
+	g_assert (sp_value_len == PROP_VALUE_MAX + 1);
+
+	gchar *env_name = g_strdup_printf ("__XA_%s", name);
+	monodroid_strreplace (env_name, '.', '_');
+	gchar *env_value = g_getenv (env_name);
+	g_free (env_name);
+
+	gsize env_value_len = env_value ? strlen (env_value) : 0;
+	if (env_value_len == 0) {
+		sp_value[0] = '\0';
+		return 0;
+	}
+
+	if (env_value_len >= sp_value_len)
+		g_warning ("System property buffer size too small by %u bytes", env_value_len == sp_value_len ? 1 : env_value_len - sp_value_len);
+
+	strncpy (sp_value, env_value, sp_value_len);
+	sp_value[sp_value_len] = '\0';
```

Comment:
Feels like an always  one byte buffer overflow.
Usually buffer size includes the null.
Perhaps not here.